### PR TITLE
Allow for graceful Exit

### PR DESCRIPTION
**Pull request by FreeMasen**
This is one possible option for allowing the invoke_handler to gracefully request a shutdown of the window to address #105.

This was the quickest thing I could think of but I am not sure it is really the best approach. One thing that might be better would be to put a new flag on `UserData` to indicate that `run` should exit

### CODE REVIEW

File: webview-sys/webview_edge.cpp [hunk 1]
```diff
@@ -591,7 +592,9 @@ WEBVIEW_API void webview_terminate(webview_t w)
 
 WEBVIEW_API void webview_exit(webview_t w)
 {
-    webview_terminate(w);
+    webview::webview* wv = static_cast<webview::webview*>(w);
+    DestroyWindow(wv->m_window);
+    OleUninitialize();
```
Comment (zxey) [inline webview-sys/webview_edge.cpp]: This should not be here, edge backend does not use OleInitialize, otherwise it seems correct.

**Comment by FreeMasen** [comment]
I believe this is ready for review. I have successfully run the new example `graceful_exit` on windows (mshtml & edge), macos and ubuntu-19.10. 

Please let me know if there is anything about the C implementations that I am misunderstanding or anything else you would prefer to see.

File: src/lib.rs [hunk 2]
```diff
@@ -431,6 +438,10 @@ impl<'a, T> WebView<'a, T> {
         }
     }
 
+    pub fn queue_close(&mut self) -> WVResult<T> {
+        Err(Error::QueueClose)
+    }
+
```
Comment (zxey) [inline src/lib.rs]: You should probably remove these, I like the idea of closing a window by returning error result, but it should probably be done in separate PR, where all the edge cases are thought through. For example this can be called outside invoke_function callback and would do nothing in that case which is weird.

File: src/lib.rs [hunk 3]
```diff
@@ -423,6 +429,7 @@ impl<'a, T> WebView<'a, T> {
                     let closure_result = &mut self.user_data_wrapper_mut().result;
                     match closure_result {
                         Ok(_) => Some(Ok(())),
+                        Err(Error::QueueClose) => None,
```
Comment (zxey) [inline src/lib.rs]: Relates to comment about queue_close.

File: src/error.rs [hunk 4]
```diff
@@ -24,6 +24,8 @@ pub enum Error {
     /// Failure to dispatch a closure to a WebView instance via a handle, likely because the
     /// WebView was dropped.
     Dispatch,
+    /// Not an error, but a graceful shutdown message from the user
+    QueueClose,
```
Comment (zxey) [inline src/error.rs]: Relates to comment about queue_close.

File: src/error.rs [hunk 5]
```diff
@@ -58,6 +60,7 @@ impl Display for Error {
                 f,
                 "Closure could not be dispatched. WebView was likely dropped."
             ),
+            Error::QueueClose => write!(f, "request to close window"),
```
Comment (zxey) [inline src/error.rs]: Relates to comment about queue_close.

File: webview-sys/webview_edge.cpp [hunk 6]
```diff
@@ -242,7 +242,8 @@ LRESULT CALLBACK WebviewWndProc(HWND hwnd, UINT msg, WPARAM wp, LPARAM lp)
         DestroyWindow(hwnd);
         break;
     case WM_DESTROY:
-        w->terminate();
+        DestroyWindow(w->m_window);
+        PostQuitMessage(0);
```
Comment (zxey) [inline webview-sys/webview_edge.cpp]: This is not correct, you are sending WM_DESTROY with DestroyWindow from webview_exit and then again here. You should remove these and replace them with w->terminate() (which you can rename to exit(), so it is consistent). w->terminate() calls browser_window::terminate() and m_webview.Close(), you can remove browser_window::terminate and keep only m_webview.Close(). There are two terminate methods: one on `webview` class and one on `browser_window` class, the one on browser_window is not needed.

File: src/lib.rs (hunk 2 above)
Comment (FreeMasen) [reply src/lib.rs]: I thought those were already removed, sorry about that

File: webview-sys/webview_edge.cpp (hunk 6 above)
Comment (FreeMasen) [reply webview-sys/webview_edge.cpp]: In my last push I have included what I believe are your suggestions. I have not had the opportunity to try and build this on a windows machine yet, I should be able to tonight.

Comment (FreeMasen) [reply webview-sys/webview_edge.cpp]: Ok, the last push is working on windows.

**Comment by zxey** [comment]
Everything looks good now, you just need to rebase your changes onto master so I can merge this.

**Comment by FreeMasen** [comment]
I was reading through some open issues and I noticed #52, I was hoping this would actually solve this but it does not appear to.

I would prefer to try and figure out how to allow for this to work correctly before merging if that is alright.

**Comment by FreeMasen** [comment]
In my last push, I included changes that would allow macos to behave (almost) the same as the other platforms.

The only caveat is that `webview_loop` will wait until the next user input is detected before return `wv->priv.should_exit`. This will actually continue to fire events in the webview (including js sourced events) that might trigger the `invoke_handler`.

**Comment by zxey** [comment]
Could the modifications to macOS exit routine alter the behavior in some way? I see that it sends message to window instead of application (is there a difference?), could it behave differently if there was for example multiple windows spawned instead of just one?

**Comment by FreeMasen** [comment]
This is still not quite finished but I think I am very close to having a correctly working solution.

Previously, telling the `NSApplication` to `terminate` is shutting down the rust process because this object is the `sharedInstance`. Sending the message to the window allows the rust process to continue but cleanup the window (which is how the gtk and windows implementations works).

Right now we are only allowing for one window by keeping the `id` of that window on the `cocoa_webview`, I haven’t messed around too much with how a user might generate another window is that possible currently? 

One option I am going to look into is using the `NSApplication. enumerateWindows` method to close all windows opened by the app. This may not be the right choice as it assumes that all windows owned by the `NSApplication` are managed by `web-view`. 

I will have one more push before I think this is ready, I am hoping I can get through it today.

**Comment by zxey** [comment]
> I haven’t messed around too much with how a user might generate another window is that possible currently?

I've created very simple example demonstrating multiple windows, you can see it on master `examples/multi_window.rs`.

**Comment by FreeMasen** [comment]
Thanks for the example, sorry I didn’t look there first. 

My expectation is that each webview instance would work independently of the other which is what we are doing in the non-macOS implementations.

**Comment by zxey** [comment]
> Thanks for the example, sorry I didn’t look there first.

No worries, I just pushed it there few minutes ago :).

> My expectation is that each webview instance would work independently of the other which is what we are doing in the non-macOS implementations.

Ok, I am a bit worried about the multiple webview example not working on macOS, because you wrote earlier:
>The only caveat is that webview_loop will wait until the next user input is detected

If one loop would wait then the other one would not progress, is that correct?

**Comment by FreeMasen** [comment]
> If one loop would wait then the other one would not progress, is that correct?

I believe I have a solution for this but with the current code yes one webview would block the other until a mouse movement or key press. In my next push I should have this fixed.

**Comment by FreeMasen** [comment]
Ok, with this last push everything is working as expected.

**Comment by FreeMasen** [comment]
As of now the branch has also been rebased onto master.

One thing I wanted to note because of your question about multiple windows, on MacOS the `multi_window` example segfaults when creating the second WebView. This happens on master meaning it is unrelated to these changes but should probably be logged as an issue.

**Comment by zxey** [comment]
Amazing work, thanks for diving so deep into this! I requested only one small change, otherwise I think this is ready to be merged. Does this also fix #52 you mentioned previously?

File: webview-sys/webview_cocoa.c [hunk 7]
```diff
@@ -88,9 +89,21 @@ static id create_menu_item(id title, const char *action, const char *key) {
...
+  // event loop.
+  id event = objc_msgSend((id)objc_getClass("NSEvent"),
+                  sel_registerName("otherEventWithType:location:modifierFlags:timestamp:windowNumber:context:subtype:data1:data2:"),
+                  NSApplicationDefinedEvent,
+                  (id)objc_getClass("NSZeroPoint"),
+                  0, 0.0, 0, NULL, 0, 0, 0);
+  id app = objc_msgSend((id)objc_getClass("NSApplication"),
+                        sel_registerName("sharedApplication"));
+  objc_msgSend(app, sel_registerName("postEvent:atStart:"), event, objc_msgSend((id)objc_getClass("NSDate"),
+                                      sel_registerName("distantPast")));
```
[the remark this answers is not in the document]
Comment (FreeMasen) [reply webview-sys/webview_cocoa.c]: Hopefully the comment I added will clear things up

**Comment by FreeMasen** [comment]
Yes, this would also fix #52
